File: koma-core/src/main/java/org/koma/core/config/KomaConfig.java
package org.koma.core.config;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.fasterxml.jackson.databind.annotation.JsonSerialize;
import org.immutables.value.Value;
import org.jetbrains.annotations.Nullable;
import org.koma.core.annotation.KomaImmutableStyle;

@Value.Immutable
@KomaImmutableStyle
@JsonSerialize(as = ImmutableKomaConfig.class)
@JsonDeserialize(as = ImmutableKomaConfig.class)
public interface KomaConfig {
  SiteConfig site();

  @Value.Default
  default OutputConfig output() {
    return ImmutableOutputConfig.builder()
      .clean(false)
      .build();
  }

  ContentConfig content();

  @Nullable
  TemplateConfig template();

  @JsonProperty("static")
  StaticConfig staticConfig();

  MetadataConfig metadata();
}

File: koma-compiler/src/main/kotlin/org/koma/core/config/KomaConfig.kt
package org.koma.core.config

import java.nio.file.Path

const val FILE_NAME = "koma.yaml"

data class ContentConfig(
  val posts: Path
)

data class MetadataConfig(
  val author: String,
  val theme: String,
  val social: Map<String, String>
)

data class OutputConfig(
  val directory: String = "dist",
  val clean: Boolean = false
)

data class ServerConfig(
  val port: Int,
)

data class SiteConfig(
  val url: String,
  val title: String,
  val description: String
)

data class TemplateConfig(
  val layout: String,
  val posts: String,
  val page: String
)

data class KomaConfig(
  val content: ContentConfig,
  val metadata: MetadataConfig,
  val output: OutputConfig,
  val server: ServerConfig,
  val site: SiteConfig
)